File: apps/backend/app/schemas/work_experience.py
"""Work experience normalization (legacy flat row → job with roles).

Normalization contract is shared with the frontend (`lib/utils/work-experience.ts`).
Keep both in sync; golden cases live in `tests/fixtures/work_experience_normalization.json`.
"""

from __future__ import annotations

import copy
from typing import Any

_TEMPLATE_PLACEHOLDER_VALUES = frozenset(
    {
        "your position title",
        "your job title",
        "position title",
        "job title",
        "your company name",
        "company name",
        "your company",
        "dates employed",
        "date range",
        "your location",
        "location",
    }
)


def normalize_experience_entry(data: Any) -> dict[str, Any]:
    """Wrap legacy flat experience rows; never merge separate entries."""
    if not isinstance(data, dict):
        return {
            "id": 0,
            "company": "",
            "location": None,
            "roles": [{"id": 1, "title": "", "years": ""}],
            "description": [],
        }

    entry = dict(data)
    roles_raw = entry.get("roles")

    if isinstance(roles_raw, list) and len(roles_raw) > 0:
        roles = [_normalize_role(role, index) for index, role in enumerate(roles_raw)]
        legacy_title = _coerce_str(entry.get("title"))
        legacy_years = _coerce_str(entry.get("years"))
        if legacy_title and not any(role.get("title") for role in roles):
            roles[0] = {**roles[0], "title": legacy_title}
        if legacy_years and not any(role.get("years") for role in roles):
            roles[0] = {**roles[0], "years": legacy_years}
    else:
        title = _coerce_str(entry.get("title"))
        years = _coerce_str(entry.get("years"))
        roles = [{"id": 1, "title": title, "years": years}]

    normalized: dict[str, Any] = {
        "id": entry.get("id", 0),
        "company": _coerce_str(entry.get("company")),
        "location": entry.get("location"),
        "roles": roles,
        "description": entry.get("description", []),
    }
    if normalized["location"] == "":
        normalized["location"] = None
    return normalized


def normalize_work_experience_list(entries: Any) -> list[dict[str, Any]]:
    """Normalize each work experience entry; preserve array length (no merging)."""
    if not isinstance(entries, list):
        return []
    return [normalize_experience_entry(entry) for entry in entries]


def preserve_work_experience_identity(
    original_entries: Any,
    tailored_entries: Any,
) -> list[dict[str, Any]]:
    """Restore employer identity from original; keep tailored descriptions.

    Used after LLM tailoring/refinement to undo flattened or split multi-role
    entries while preserving edited bullet text.
    """
    if not isinstance(original_entries, list):
        return normalize_work_experience_list(tailored_entries)

    tailored_list = tailored_entries if isinstance(tailored_entries, list) else []
    preserved: list[dict[str, Any]] = []

    for index, orig_entry in enumerate(original_entries):
        if not isinstance(orig_entry, dict):
            continue

        orig_normalized = normalize_experience_entry(orig_entry)
        tailored_entry = (
            tailored_list[index] if index < len(tailored_list) else None
        )
        if isinstance(tailored_entry, dict) and "description" in tailored_entry:
            desc = tailored_entry["description"]
            if isinstance(desc, list):
                tailored_description = copy.deepcopy(desc)
            elif desc is None:
                tailored_description = []
            else:
                tailored_description = [copy.deepcopy(desc)]
        else:
            # Missing tailored row or no description field — keep master/original pool.
            tailored_description = copy.deepcopy(
                orig_normalized.get("description", [])
            )

        preserved.append(
            {
                "id": orig_normalized.get("id", 0),
                "company": orig_normalized.get("company", ""),
                "location": orig_normalized.get("location"),
                "roles": copy.deepcopy(orig_normalized.get("roles", [])),
                "description": tailored_description,
            }
        )

    return preserved


def _normalize_role(role: Any, index: int) -> dict[str, Any]:
    if not isinstance(role, dict):
        return {"id": index + 1, "title": "", "years": ""}
    role_id = role.get("id", index + 1)
    return {
        "id": role_id if isinstance(role_id, int) else index + 1,
        "title": _coerce_str(role.get("title")),
        "years": _coerce_str(role.get("years")),
    }


def _coerce_str(value: Any) -> str:
    if value is None:
        return ""
    text = str(value)
    if text.strip().casefold() in _TEMPLATE_PLACEHOLDER_VALUES:
        return ""
    return text


def experience_role_titles(entry: dict[str, Any]) -> str:
    """Joined role titles for labels; falls back to legacy top-level title."""
    roles = entry.get("roles")
    titles: list[str] = []
    if isinstance(roles, list):
        for role in roles:
            if isinstance(role, dict):
                title = _coerce_str(role.get("title"))
                if title:
                    titles.append(title)
    if not titles:
        legacy = _coerce_str(entry.get("title"))
        if legacy:
            titles.append(legacy)
    return "; ".join(titles)


def roles_identity_pairs(entry: dict[str, Any]) -> list[tuple[str, str]]:
    """Stable (title, years) pairs for identity comparison."""
    normalized = normalize_experience_entry(entry)
    pairs: list[tuple[str, str]] = []
    for role in normalized.get("roles", []):
        if isinstance(role, dict):
            pairs.append(
                (_coerce_str(role.get("title")), _coerce_str(role.get("years")))
            )
    return pairs


def compare_work_experience_identity(
    orig: dict[str, Any], res: dict[str, Any], index: int
) -> list[str]:
    """Return warnings when company or role title/years change."""
    warnings: list[str] = []

    orig_company = _coerce_str(orig.get("company"))
    res_company = _coerce_str(res.get("company"))
    if orig_company and orig_company != res_company:
        warnings.append(
            f"Identity field changed: workExperience[{index}].company "
            f"('{orig_company}' → '{res_company}')"
        )

    orig_roles = roles_identity_pairs(orig)
    res_roles = roles_identity_pairs(res)
    if len(orig_roles) != len(res_roles):
        warnings.append(
            f"Identity field changed: workExperience[{index}].roles "
            f"(count {len(orig_roles)} → {len(res_roles)})"
        )

    for role_idx, (orig_pair, res_pair) in enumerate(zip(orig_roles, res_roles)):
        for field_name, orig_val, res_val in (
            ("title", orig_pair[0], res_pair[0]),
            ("years", orig_pair[1], res_pair[1]),
        ):
            if orig_val and orig_val != res_val:
                warnings.append(
                    f"Identity field changed: workExperience[{index}].roles[{role_idx}].{field_name} "
                    f"('{orig_val}' → '{res_val}')"
                )

    return warnings

File: apps/frontend/lib/utils/work-experience.ts
import type { Experience, ExperienceRole } from '@/components/dashboard/resume-component';

/** Template filler text from Word/PDF forms — treat as empty so placeholders show instead.
 * Normalization contract is shared with the backend (`app/schemas/work_experience.py`).
 * Keep both in sync; golden cases live in `tests/fixtures/work_experience_normalization.json`.
 */
const TEMPLATE_PLACEHOLDER_VALUES = new Set(
  [
    'your position title',
    'your job title',
    'position title',
    'job title',
    'your company name',
    'company name',
    'your company',
    'dates employed',
    'date range',
    'your location',
    'location',
  ].map((s) => s.toLowerCase())
);

function coerceStr(value: unknown): string {
  if (value == null) return '';
  const text = String(value);
  if (TEMPLATE_PLACEHOLDER_VALUES.has(text.trim().toLowerCase())) {
    return '';
  }
  return text;
}

function normalizeRole(role: unknown, index: number): ExperienceRole {
  if (!role || typeof role !== 'object') {
    return { id: index + 1, title: '', years: '' };
  }
  const r = role as Record<string, unknown>;
  const roleId = typeof r.id === 'number' ? r.id : index + 1;
  return {
    id: roleId,
    title: coerceStr(r.title),
    years: coerceStr(r.years),
  };
}

/** Wrap legacy flat experience rows; never merge separate entries. */
export function normalizeExperienceEntry(data: unknown): Experience {
  if (!data || typeof data !== 'object') {
    return {
      id: 0,
      company: '',
      location: '',
      roles: [{ id: 1, title: '', years: '' }],
      description: [],
    };
  }

  const entry = data as Record<string, unknown>;
  const rolesRaw = entry.roles;

  let roles: ExperienceRole[];
  if (Array.isArray(rolesRaw) && rolesRaw.length > 0) {
    roles = rolesRaw.map((role, index) => normalizeRole(role, index));
    const legacyTitle = coerceStr(entry.title);
    const legacyYears = coerceStr(entry.years);
    if (legacyTitle && !roles.some((role) => role.title)) {
      roles[0] = { ...roles[0], title: legacyTitle };
    }
    if (legacyYears && !roles.some((role) => role.years)) {
      roles[0] = { ...roles[0], years: legacyYears };
    }
  } else {
    roles = [
      {
        id: 1,
        title: coerceStr(entry.title),
        years: coerceStr(entry.years),
      },
    ];
  }

  const description = Array.isArray(entry.description)
    ? entry.description.map((d) => (typeof d === 'string' ? d : String(d ?? '')))
    : [];

  return {
    id: typeof entry.id === 'number' ? entry.id : 0,
    company: coerceStr(entry.company),
    location: coerceStr(entry.location) || undefined,
    roles,
    description,
  };
}

export function normalizeWorkExperience(entries: unknown): Experience[] {
  if (!Array.isArray(entries)) return [];
  return entries.map((entry) => normalizeExperienceEntry(entry));
}

/** Ensure each job has a roles array without stripping user input (for live editing). */
export function ensureExperienceRoles(entries: Experience[]): Experience[] {
  return entries.map((entry) => {
    if (entry.roles?.length) {
      return entry;
    }
    const legacy = entry as Experience & { title?: string; years?: string };
    return {
      ...entry,
      roles: [{ id: 1, title: legacy.title ?? '', years: legacy.years ?? '' }],
    };
  });
}

export function withNormalizedWorkExperience<T extends { workExperience?: unknown }>(
  data: T
): T & { workExperience: Experience[] } {
  return {
    ...data,
    workExperience: normalizeWorkExperience(data.workExperience),
  };
}

/** Primary role title for labels (first role). */
export function getExperiencePrimaryTitle(job: Experience): string {
  return job.roles?.[0]?.title ?? '';
}

/** All role titles joined for display. */
export function getExperienceRoleTitles(job: Experience): string {
  return (job.roles ?? [])
    .map((r) => r.title?.trim())
    .filter(Boolean)
    .join('; ');
}

/** Combined years line across roles (for compact single-line templates). */
export function getExperienceYearsLine(job: Experience): string {
  return (job.roles ?? [])
    .map((r) => r.years?.trim())
    .filter(Boolean)
    .join(' · ');
}
